File: cli-validation/python-cli/src/documentation_robotics/commands/project.py
"""
Project an element to other layers.
"""

from pathlib import Path
from typing import Optional

import click
from rich.console import Console
from rich.table import Table

from ..core.model import Model

console = Console()


@click.command()
@click.argument("element-id")
@click.option("--to", "to_layers", required=True, help="Comma-separated target layers")
@click.option("--rule", help="Specific projection rule to use")
@click.option("--dry-run", is_flag=True, help="Show what would be created")
@click.option("--force", is_flag=True, help="Overwrite existing elements")
def project(element_id: str, to_layers: str, rule: Optional[str], dry_run: bool, force: bool):
    """Project an element to other layers."""

    # Load model
    try:
        model = Model(Path.cwd())
    except FileNotFoundError:
        console.print("✗ Error: No model found in current directory", style="red bold")
        raise click.Abort()

    # Get source element
    source = model.get_element(element_id)
    if not source:
        console.print(f"✗ Error: Element '{element_id}' not found", style="red bold")
        raise click.Abort()

    # Initialize projection engine (already initialized in model)
    engine = model.projection_engine

    # Parse target layers
    target_layers = [layer.strip() for layer in to_layers.split(",")]

    console.print(f"\n[bold]Projecting element:[/bold] {element_id}")
    console.print(f"[bold]Target layers:[/bold] {', '.join(target_layers)}")

    # Project to each layer
    results = []

    for target_layer in target_layers:
        # Validate target layer
        if not model.get_layer(target_layer):
            console.print(f"⚠  Warning: Layer '{target_layer}' not found, skipping", style="yellow")
            continue

        try:
            # Find applicable rule
            applicable_rules = engine.find_applicable_rules(source, target_layer)

            if not applicable_rules:
                console.print(
                    f"⚠  Warning: No projection rule found for {source.layer} -> {target_layer}",
                    style="yellow",
                )
                continue

            projection_rule = applicable_rules[0]

            # Perform projection
            projected = engine.project_element(
                source, target_layer, rule=projection_rule, dry_run=dry_run
            )

            if projected:
                results.append((target_layer, projected))

        except Exception as e:
            console.print(f"✗ Error projecting to {target_layer}: {e}", style="red")

    # Display results
    if results:
        console.print("\n[bold green]✓ Projection successful[/bold green]\n")

        table = Table(title="Projected Elements")
        table.add_column("Target Layer", style="cyan")
        table.add_column("Element ID", style="white")
        table.add_column("Name", style="green")

        for target_layer, element in results:
            table.add_row(target_layer, element.id, element.name)

        console.print(table)

        if dry_run:
            console.print("\n[yellow]Dry run - elements not saved[/yellow]")
        else:
            # Save model
            model.save()
            console.print("\n[green]Model saved successfully[/green]")
    else:
        console.print("\n[yellow]No elements were projected[/yellow]")


@click.command()
@click.option("--from", "from_layer", help="Source layer filter")
@click.option("--to", "to_layer", help="Target layer filter")
@click.option("--dry-run", is_flag=True, help="Show what would be created")
def project_all(from_layer: Optional[str], to_layer: Optional[str], dry_run: bool):
    """Project all applicable elements."""

    # Load model
    try:
        model = Model(Path.cwd())
    except FileNotFoundError:
        console.print("✗ Error: No model found in current directory", style="red bold")
        raise click.Abort()

    # Initialize projection engine
    engine = model.projection_engine

    console.print("[bold]Projecting all applicable elements...[/bold]\n")

    if from_layer:
        console.print(f"From layer: {from_layer}")
    if to_layer:
        console.print(f"To layer: {to_layer}")

    # Perform projections
    projected = engine.project_all(from_layer, to_layer, dry_run)

    # Display results
    if projected:
        console.print(f"\n[bold green]✓ Projected {len(projected)} element(s)[/bold green]\n")

        # Group by layer
        by_layer = {}
        for element in projected:
            if element.layer not in by_layer:
                by_layer[element.layer] = []
            by_layer[element.layer].append(element)

        for layer, elements in sorted(by_layer.items()):
            console.print(f"[bold]{layer}:[/bold]")
            for element in elements:
                console.print(f"  • {element.id} ({element.name})")

        if dry_run:
            console.print("\n[yellow]Dry run - elements not saved[/yellow]")
        else:
            # Save model
            model.save()
            console.print("\n[green]Model saved successfully[/green]")
    else:
        console.print("\n[yellow]No elements were projected[/yellow]")

File: cli/src/commands/project.ts
/**
 * Project command - project elements across layers using projection rules
 */

import ansis from "ansis";
import { Model } from "../core/model.js";

export async function projectCommand(
  elementId: string,
  targetLayers: string,
  options: {
    rule?: string;
    dryRun?: boolean;
    force?: boolean;
    model?: string;
  } = {}
): Promise<void> {
  try {
    // Load model
    const model = await Model.load(options.model);

    // Get source element
    const source = model.getElementById(elementId);
    if (!source) {
      console.error(ansis.red(`Error: Element '${elementId}' not found`));
      process.exit(1);
    }

    // Initialize projection engine
    const engine = model.getProjectionEngine();

    // Parse target layers
    const layers = targetLayers.split(",").map((l) => l.trim());

    console.log("");
    console.log(ansis.bold(`Projecting element: ${ansis.yellow(elementId)}`));
    console.log(ansis.bold(`Target layers: ${ansis.cyan(layers.join(", "))}`));
    console.log("");

    // Project to each layer
    const results: Array<{ layer: string; element: any }> = [];

    for (const targetLayer of layers) {
      try {
        // Find applicable rules
        const applicableRules = engine.findApplicableRules(source, targetLayer);

        if (applicableRules.length === 0) {
          console.log(
            ansis.yellow(
              `⚠  Warning: No projection rule found for ${source.layer} -> ${targetLayer}`
            )
          );
          continue;
        }

        // Use first applicable rule (or specific rule if provided)
        const projectionRule = options.rule
          ? applicableRules.find((r) => r.name === options.rule)
          : applicableRules[0];

        if (!projectionRule) {
          console.log(
            ansis.yellow(`⚠  Warning: Rule '${options.rule}' not found for ${targetLayer}`)
          );
          continue;
        }

        // Perform projection
        const projected = await engine.projectElement(
          source,
          targetLayer,
          projectionRule,
          options.dryRun
        );

        if (projected) {
          results.push({ layer: targetLayer, element: projected });
        }
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        console.log(ansis.red(`✗ Error projecting to ${targetLayer}: ${message}`));
      }
    }

    // Display results
    if (results.length > 0) {
      console.log(ansis.bold.green("✓ Projection successful"));
      console.log("");

      // Table header
      const colLayer = "Target Layer";
      const colId = "Element ID";
      const colName = "Name";

      const maxLayerLen = Math.max(colLayer.length, ...results.map((r) => r.layer.length));
      const maxIdLen = Math.max(colId.length, ...results.map((r) => r.element.id.length));
      const maxNameLen = Math.max(
        colName.length,
        ...results.map((r) => (r.element.name || "").length)
      );

      // Print header
      console.log(
        ansis.bold(
          ansis.cyan(colLayer.padEnd(maxLayerLen)) +
            "  " +
            ansis.white(colId.padEnd(maxIdLen)) +
            "  " +
            ansis.green(colName.padEnd(maxNameLen))
        )
      );
      console.log(ansis.dim("─".repeat(maxLayerLen + maxIdLen + maxNameLen + 4)));

      // Print rows
      for (const { layer, element } of results) {
        console.log(
          ansis.cyan(layer.padEnd(maxLayerLen)) +
            "  " +
            ansis.white(element.id.padEnd(maxIdLen)) +
            "  " +
            ansis.green((element.name || "").padEnd(maxNameLen))
        );
      }

      console.log("");

      if (options.dryRun) {
        console.log(ansis.yellow("Dry run - elements not saved"));
      } else {
        // Save model
        await model.save();
        console.log(ansis.green("Model saved successfully"));
      }
    } else {
      console.log(ansis.yellow("No elements were projected"));
    }

    console.log("");
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(ansis.red(`Error: ${message}`));
    process.exit(1);
  }
}

export async function projectAllCommand(
  options: {
    from?: string;
    to?: string;
    dryRun?: boolean;
    model?: string;
  } = {}
): Promise<void> {
  try {
    // Load model
    const model = await Model.load(options.model);

    // Initialize projection engine
    const engine = model.getProjectionEngine();

    console.log("");
    console.log(ansis.bold("Projecting all applicable elements..."));
    console.log("");

    if (options.from) {
      console.log(`From layer: ${options.from}`);
    }
    if (options.to) {
      console.log(`To layer: ${options.to}`);
    }
    if (options.from || options.to) {
      console.log("");
    }

    // Perform projections
    const projected = await engine.projectAll(
      options.from || undefined,
      options.to || undefined,
      options.dryRun
    );

    // Display results
    if (projected.length > 0) {
      console.log(ansis.bold.green(`✓ Projected ${projected.length} element(s)`));
      console.log("");

      // Group by layer
      const byLayer = new Map<string, any[]>();
      for (const element of projected) {
        const layer = element.layer || "unknown";
        if (!byLayer.has(layer)) {
          byLayer.set(layer, []);
        }
        byLayer.get(layer)!.push(element);
      }

      // Display grouped results
      for (const [layer, elements] of Array.from(byLayer.entries()).sort()) {
        console.log(ansis.bold(`${layer}:`));
        for (const element of elements) {
          console.log(
            `  ${ansis.dim("•")} ${element.id} ${ansis.gray(`(${element.name || "unnamed"})`)}`
          );
        }
        console.log("");
      }

      if (options.dryRun) {
        console.log(ansis.yellow("Dry run - elements not saved"));
      } else {
        // Save model
        await model.save();
        console.log(ansis.green("Model saved successfully"));
      }
    } else {
      console.log(ansis.yellow("No elements were projected"));
    }

    console.log("");
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(ansis.red(`Error: ${message}`));
    process.exit(1);
  }
}
